### PR TITLE
Asynchronously prepare next segment to reduce peak write latency

### LINKED ISSUE #11444: Asynchronously prepare next segment to reduce peak write latency
In order to avoid spikes in our p95 and p99 latency due to segment creation, we should prototype creating (and pre-allocating) the next segment asynchronously. 

We can do this at various points in time. The easiest option is probably to trigger pre-creation whenever we need to create a new segment. Another, more advanced option, is to start pre-creating a segment when the writer has filled x% of the current segment.

# Support
- SUPPORT-15889

### PR DESCRIPTION
**Pull request by lenaschoenburg**
Whenever a new segment acquired for a writer, we trigger an async task to create the next needed segment ahead of time. The next time a writer needs a new segment, the file is already created and filled by the segment allocator. This reduces peak write latency because the writer thread only has to wait for task completion without doing any expensive IO.

The next segment is created uninitialized, without a segment descriptor. On first access of the next segment, we initialize it by writing the correct descriptor. When an uninitialized segment exists on recovery, it is detected as a corrupted log segment (because it does not have the expected descriptor) and thus deleted.

The segment id used to be a monotonically increasing number. Since we increase this number before preparing the next segment, any log resets may break this assumption of monotonically increasing ids.
To illustrate, lets look at an example: the log may have log segments 1-3, and an uninitialized segment 4. When the log is then reset before initializing and writing to 4, log  segments 2 and 3 are deleted while 4 is not indexed and thus not affected by the reset. When continuing to write to the log, segment 4 will be used, resulting in a consistent state of the journal but with segments 1 and 4. This may be undesirable for troubleshooting but causes no other issues.

p99 write latency with async prepare:
image

p99 write latency without async prepare: 
image

Without async preparation of the next segment, write latency can peak at ~ 2 seconds when the write creates a new segment.

closes #11444

### CODE REVIEW

**Comment by npepinpe** [comment]
I'd be curious to see how this performs under heavy load. I imagine if write i/o is a bottleneck, we would still see a spike due to doing some i/o. That said, I think for most workloads, we don't max out the disk throughput (though we do max out iops), so maybe we can wait until write i/o really is a bottleneck before looking into file recycling (which has little to no i/o penalty), and even combine both approaches in the future.

**Comment by oleschoenburg** [comment]
I think it should be possible and safe to backport this, wdyt @deepthidevaki?

**Comment by deepthidevaki** [comment]
> even combine both approaches in the future.

Combining both would be also useful, I think. I/O from async creation can influence other concurrent i/o also from other partitions. So overall reducing any i/o would be beneficial. Currently, with async create we saw some spikes in segment creation time. But I guess it is coming from the flush of previous segment.

**Comment by deepthidevaki** [comment]
> I think it should be possible and safe to backport this, wdyt @deepthidevaki?

It should be possible. But I'm not sure if we should backport this. I wouldn't do it by default as it is a new feature. We can do it if it is necessary to reach the target performance in 8.1.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java [hunk 1]
```diff
@@ -48,7 +48,7 @@ final class DiskSpaceRecoveryIT {
           .withEnv("ZEEBE_BROKER_EXPERIMENTAL_RAFT_PREFERSNAPSHOTREPLICATIONTHRESHOLD", "0")
           .withEnv("ZEEBE_BROKER_DATA_LOGSEGMENTSIZE", "1MB")
           .withEnv("ZEEBE_BROKER_NETWORK_MAXMESSAGESIZE", "1MB")
-          .withEnv("ZEEBE_BROKER_DATA_DISK_FREESPACE_PROCESSING", "10MB")
+          .withEnv("ZEEBE_BROKER_DATA_DISK_FREESPACE_PROCESSING", "8MB")
```
Comment (deepthidevaki) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java:51]: :question: Why was this required?

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
Thanks @oleschoenburg for taking up this.

:x: I missed one thing when we discussed this. Free diskspace check is not optimal with this change. You would have created an uninitialized segment. When the writer tries gets the next segment, it sees that free disk space is below the threshold and return OOD, even though there is an empty segment which it can use. This must be fixed. Probably, we can move the check to more close to where the file is allocated. Let's sync on this before merging this PR.

**Comment by oleschoenburg** [comment]
@deepthidevaki good point regarding disk size, I completely missed that.
While trying to refactor this a bit I stumbled upon `SEGMENT_BUFFER_FACTOR=3`. Am I understanding correctly that the journal requires space for at least three segments? Do you remember what the reasoning behind that was and if it still applies?
I don't see any big issues with just requiring enough space for one segment when we know that we'll only create one. The minFreeSpace config is responsible for leaving enough free space for all other disk usage.

**Comment by deepthidevaki** [comment]
I guess it is a left over from atomix code. I don't see any reason to keep space for 3 segments. As you said it is enough to have space for one segment.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java (hunk 1 above)
Comment (oleschoenburg) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/health/DiskSpaceRecoveryIT.java:51]: The test times out otherwise. I can't really wrap my head around it (for example: how much disk space is from segments, how much from RocksDB) and it's a but unsatisfying to just change a magic number but currently I don't see a better option.

The only difference here is that we'll always have at least two segments (each 1MiB in this test) instead of just one.

**Review by deepthidevaki — APPROVED** [review]
Great work :rocket: 

:wrench: It might be useful to document that id in the segment file name is not same as actual segment id in the descriptor. `SegmentFile::getSegmentIdFromPath` looks like a good place to document this so that we don't accidentally use that in future.

**Comment by deepthidevaki** [comment]
@oleschoenburg Please also link the issue #11444 to the PR description and update the title of the issue.
